src/itwinai/torch/distributed.py

```diff
@@ -1010,3 +1007,180 @@ def gather(self, tensor: torch.Tensor, dst_rank: int = 0):
+                    )
+
+            except Exception as e:
+                raise RuntimeError(f"Error initializing Ray: {str(e)}")
+
+            ray.init(address="auto")
+
+
+class RayDDPStrategy(RayDistributedStrategy):
+    def __init__(self) -> None:
+        super()._initialize_ray()
+
+    def init(self) -> None:
+        self.is_initialized = True
+
+    @check_initialized
+    def global_world_size(self) -> int:
+        return ray.train.get_context().get_world_size()
+
+    @check_initialized
+    def local_world_size(self) -> int:
+        return ray.train.get_context().get_local_world_size()
+
+    @check_initialized
+    def global_rank(self) -> int:
+        return ray.train.get_context().get_world_rank()
+
+    @check_initialized
+    def local_rank(self) -> int:
+        return ray.train.get_context().get_local_rank()
+
+    @check_initialized
+    def distributed(
+        self,
+        model: nn.Module,
+        optimizer: Optimizer,
+        lr_scheduler: Optional[LRScheduler] = None,
+    ) -> Tuple[Module, Optimizer, LRScheduler | None]:
+        model = ray.train.torch.prepare_model(model)
+
+        return model, optimizer, lr_scheduler
+
+    @check_initialized
+    def create_dataloader(
+        self,
+        dataset: Dataset[T_co],
+        batch_size: Optional[int] = 1,
+        shuffle: Optional[bool] = None,
+        sampler: Union[Sampler, Iterable, None] = None,
+        batch_sampler: Union[Sampler[List], Iterable[List], None] = None,
+        num_workers: int = 0,
+        collate_fn: Optional[_collate_fn_t] = None,
+        pin_memory: bool = False,
+        drop_last: bool = False,
+        timeout: float = 0,
+        worker_init_fn: Optional[_worker_init_fn_t] = None,
+        multiprocessing_context=None,
+        generator=None,
+        *,
+        prefetch_factor: Optional[int] = None,
+        persistent_workers: bool = False,
+        pin_memory_device: str = "",
+    ):
+        dataloader = super().create_dataloader(
+            dataset=dataset,
+            batch_size=batch_size,
+            sampler=sampler,
+            shuffle=shuffle,
+            num_workers=num_workers,
+            collate_fn=collate_fn,
+            pin_memory=pin_memory,
+            drop_last=drop_last,
+            timeout=timeout,
+            worker_init_fn=worker_init_fn,
+            multiprocessing_context=multiprocessing_context,
+            generator=generator,
+            prefetch_factor=prefetch_factor,
+            persistent_workers=persistent_workers,
+            pin_memory_device=pin_memory_device,
+        )
+
+        return ray.train.torch.prepare_data_loader(dataloader)
+
+    def clean_up(self) -> None:
+        pass
+
+    def allgather_obj(self, obj: Any) -> List[Any]:
+        pass
+
+    def gather_obj(self, obj: Any, dst_rank: int = 0) -> List[Any]:
+        pass
+
+    def gather(self, tensor: torch.Tensor, dst_rank: int = 0) -> List | None:
+        pass
```

Comment:
This is one of the things that I wanted to look at in a future PR, I had really not thought about these methods and because so far in the virgo integration they haven't come up I thought it would be okay to leave them like this for now?

Reply:
I think this is a bit too critical part to skip before merging, because developing it may change a lot the structure of the Ray strategies

Reply:
Okay I changed the RayDDPStrategy that it now inherits from the TorchDDPStrategy - I think the way the strategies are organised now is the best way to do it, and it turns out, `gather`, `allgather_obj` and `gather_obj` could just be reused from their parent strategies (I have tested them in both DDP and DeepSpeed and both work).

Reply:
I had looked at the ray collective library but I believe that this is geared more towards usage with ray core and not needed when we are using the torch distributed backend (which is great, because it would complicate our code so much 🙌🏼)